### ISSUE TITLE
Unread messages implement in the server

### ISSUE THREAD

**Issue by delijati** [matrix-org/synapse#2632]
This is way to much code to be implemented in every client [1]

[1] https://github.com/matrix-org/matrix-python-sdk/issues/161

**Comment by ara4n** [comment]
I agree that the balance is wrong, and irritatingly I'm failing to remember how we've ended up like this.  Just to confirm the current situation:
 * The server calculates the number of **unread highlighted notifications** for a room (the `unread_notifications.highlight_count` field of the /sync response)
 * It also calculates the number of **unread notifications** in general (the `unread_notifications.notification-count` field.

You're right that it doesn't track the number of unread messages in general though, and this has to be inferred by the client by looping through all the messages which the client has received in a room since we last sent a Read Receipt, and summing them up.

Possible reasons for this are:
 1. It's quite a heavy operation for the server to do, so we offloaded it to the client (whereas doing it for the subset of messages which are *notifications* is less heavy as there's less data)
 2. We're forced into calculating the unread notification counts serverside anyway to update push notification badges
 3. Clients can apply whatever weird filtering rules they like to show given events, so in the end it's really up to the client to decide whether a given message should actually nudge the unread count for a room or not.  (e.g. should an `m.call.candidates` event in a room nudge the unread count or not?).

In practice, calculating the unread counts clientside is still a trainwreck, as evinced by https://github.com/vector-im/riot-web/issues/3363, so we really do need to find a way to fix this.

Perhaps one solution would be to define stricter rules on the events which should constitute an "unread messages" in the spec, and thus have that metric be consistent across all clients.  This could in turn tie into the long-awaited "human readable event representation" work (https://github.com/matrix-org/matrix-doc/pull/92 and its many descendants), as in the end the concept of "unreadness" is very much a human representation question.  The same metric that we use to decide that a message is human readable could then also be used to calculate unreadness serverside, and everyone wins.

**Comment by komzbike** [comment]
+1

**Comment by nunoperalta** [comment]
This is being discussed here for Riot:
https://github.com/vector-im/riot-web/issues/4406

The only blocker at the moment for Riot is that synapse only ever let's the "unread" pointer move forward.

This needs an MSC so that it's in the spec and not specific to riot/synapse.

Slack, Facebook Messenger, WhatsApp, Skype, Linked In, Signal, and pretty much every other email/chat client I know, have Mark as Unread, and some of them (like Skype and Slack) allow to choose which message to move the unread pointer to.

Thank you.

**Comment by nunoperalta** [comment]
Related conversation #5601

**Comment by nunoperalta** [comment]
Any update on this?
This has been asked already since 2017, in #4406 ...
This is one of the most basic features a Chat system like this needs to have.
Thank you.

**Comment by richvdh** [comment]
this is a different problem to #5601, and #4406 is *completely* unrelated.

**Comment by richvdh** [comment]
this is related to https://github.com/matrix-org/matrix-doc/pull/2654, though I don't think it will do much to change the situation here.

**Comment by nunoperalta** [comment]
Apologies - I meant https://github.com/vector-im/riot-web/issues/4406
Different project :)

**Comment by clokep** [comment]
Note that MSC2654 was implemented in Synapse in #8059, albeit behind an experimental flag.
